### DISCUSSION TITLE
Phoneset

### DISCUSSION THREAD

**Discussion by elnursamedzade** [RHVoice/RHVoice#481] [Q&A]
Hello.
I have a few questions.
1. Can I use my alignment (.textgrid) files that I created with Montreal forced aligner for RHVoice?
2. From where does RHVoice read phoneset data?
3. I want to make my own alophone set by editing the phoneset file. How do I do this?
Thank you.

**Comment by Olga-Yakovleva** [comment]
1. RHVoice uses htk-format labels. So just write a custom script to 
convert whatever format your labels use and skip the segmentation step 
when you train your voice. The g2p rules you implement in RHVoice for 
your language need to output the same transcriptions though, except for 
pauses.

2 and 3. See the phonemes.xml files in existing languages, such as 
data/languages/English/phonemes.xml etc. You need to define a similar 
phoneme set with phonemes and features appropriate for your language.

**Reply by elnursamedzade** [reply]
Hello Olga. Thank you for taking your precious time to reply.
Yes, I wrote a textgrid to lab script.
I placed the lab files in the data/labels/mono directory.
I have my own transcriptions in my prompts.ssml file.
When I run:
../RHVoice/src/scripts/general/voice-building-utils label
I am getting this output:

data/labels/mono/RHVoice_elnur_0001.lab
Label mismatch: expected ow, found O
Traceback (most recent call last):
  File "../RHVoice/src/scripts/general/voice-building-utils", line 1613, in <module>
    args.func(args)
  File "../RHVoice/src/scripts/general/voice-building-utils", line 627, in __call__
    subprocess.check_call([os.path.join(RHVoice_bindir,"RHVoice-make-hts-labels"),"-l",os.path.join("data","labels","mono"),self.settings["text"],labdir])
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/elka/tts/RHVoice/local/bin/RHVoice-make-hts-labels', '-l', 'data/labels/mono', '/home/elka/tts/elnur/prompts.ssml', 'data/labels/full']' returned non-zero exit status 255.

What should I do?

**Comment by Olga-Yakovleva** [comment]
What language are you trying to build a voice for? Are you working on 
support for a new language? Have you implemented pronunciation rules for 
it using the RHVoice's native conventions, which you need to do if you 
want to build a voice for RHVoice? If you are trying to build a voice 
for a language RHVoice already supports, such as american English, you 
need to make sure that your alignments match what RHVoice's 
transcription rules generate exactly. You need to map your source 
phonemes to the native ones of RHVoice, and the sequence of segments 
needs to match exactly in every sentence. Btw, only lower case letters 
are supported in the names of phonemes in RHVoice. No IPA or X-SAMPA, 
such names need to be mapped.

Olga

elnursamedzade wrote:

**Comment by Olga-Yakovleva** [comment]
Actually, provided that the languages properly implemented in RHVoice, 
you could also try creating an ssml using the phoneme tag to transcribe 
every word, again, using native phonemes. This is a recent feature 
though, and may contain bugs I haven't found yet. Most likely does 
contain bugs.

**Comment by elnursamedzade** [comment]
Hello Olga.
I solved the above problem. It was due to my mistakes with phoneset. Thank you.

I am trying to make language for Azerbaijani language.
I am currently converting the foma data to RHVoice fst format by editing it for my language.

Some questions arise...
1. What do I need to do for Gpos.foma? For example, I'm looking at the Georgian sample file. There are some words there. What do these mean?
2. What is Labeling.xml used for? How should I edit it for my language?
3. Spell.foma, syl.foma, tok.foma - what are these files for?

**Comment by zstanecic** [comment]
Hi Elnur!

Gpos means general parts of speech, there are prepositions, postpositions, pronouns, and eventually adverbs.

You can check the polish language example.

Tok is the tokenizer,

Syl is the syllabification.

Spelling is used to define how symbols and numbers with letters should be spelled

**Reply by elnursamedzade** [reply]
Hello @zstanecic.
Yes I understand. What does the Labelling.xml file do?

**Comment by Olga-Yakovleva** [comment]
Labeling.xml defines the format of full-context labels used for training 
and synthesis. These labels encode the linguistic specification as a 
sequence of feature values separated by special characters for ease of 
matching. Please see this paper for a good explanation of what I mean by 
a linguistic specification and how the statistical parametric synthesis 
works in general from training to runtime speech generation.

https://era.ed.ac.uk/handle/1842/4541

**Reply by elnursamedzade** [reply]
Hi @Olga.
Does the following output have anything to do with labelling.xml? :
Label mismatch: expected u, found a

**Comment by Olga-Yakovleva** [comment]
It means that your aligned labels don't match the transcription 
generated by RHVoice, the situation I warned about in one of my previous 
replies. I'm assuming because you aren't using RHVoice's native 
alignment implementation. I think you should let RHVoice transcribe and 
align, at least to start with. This way you can learn the whole process 
without this additional complication of an external aligner. Later on, 
if you think that external aligner generates more accurate boundaries, 
which may be the case, especially if it uses a good language-specific 
speech recognition model trained on many speakers, you can try it. But 
you will need to arrange it so that the two transcriptions match 
exactly. But you are implementing your pronunciation rules, so you have 
full control of what RHVoice outputs.

**Reply by elnursamedzade** [reply]
But it is progressing a little.
0001.lab 0002.lab 0003.lab ...
I leave the full output below:
elka@musician:~/tts/elnur$ ../RHVoice/src/scripts/general/voice-building-utils label
data/labels/mono/RHVoice_elnur_0001.lab
data/labels/mono/RHVoice_elnur_0002.lab
data/labels/mono/RHVoice_elnur_0003.lab
data/labels/mono/RHVoice_elnur_0004.lab
data/labels/mono/RHVoice_elnur_0005.lab
data/labels/mono/RHVoice_elnur_0006.lab
data/labels/mono/RHVoice_elnur_0007.lab
data/labels/mono/RHVoice_elnur_0008.lab
data/labels/mono/RHVoice_elnur_0009.lab
data/labels/mono/RHVoice_elnur_0010.lab
data/labels/mono/RHVoice_elnur_0011.lab
Label mismatch: expected e, found a
Traceback (most recent call last):
  File "../RHVoice/src/scripts/general/voice-building-utils", line 1613, in <module>
    args.func(args)
  File "../RHVoice/src/scripts/general/voice-building-utils", line 627, in __call__
    subprocess.check_call([os.path.join(RHVoice_bindir,"RHVoice-make-hts-labels"),"-l",os.path.join("data","labels","mono"),self.settings["text"],labdir])
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/elka/tts/RHVoice/local/bin/RHVoice-make-hts-labels', '-l', 'data/labels/mono', '/home/elka/tts/elnur/prompts.ssml', 'data/labels/full']' returned non-zero exit s
tatus 255.

I have 10 full context lab files in the data/full directory. When I examine the Mono/RHVoice_elnur_0011.lab file, I don't see any errors.
I set the language data according to my language. I defined my own alophon set in the phoneset.xml file.

**Comment by Olga-Yakovleva** [comment]
Please run RHVoice-transcribe on your ssml and compare the corresponding 
line in the output.

**Comment by Olga-Yakovleva** [comment]
RHVoice-transcribe-sentenses <ssml> out.txt

**Reply by elnursamedzade** [reply]
I made it! 😂😂
I rerun the segmentation phase and this time it completed miraculously without errors.
I have HTK formatted lab files in my Labels/mono dir.
I have also completed the full context lab phase and am currently in the training process.
Thank you @Olga! You are really doing great things for us. My best wishes and love.
